### PR TITLE
handling realm in /json requests

### PR DESCRIPTION
**Pull request by mdetko**
Builds realm path according to the doc: https://backstage.forgerock.com/docs/am/6/dev-guide/#sec-rest-realms

### CODE REVIEW

File: lib/openam.js [hunk 1]
```diff
@@ -73,16 +72,22 @@ OpenAMClient.prototype.authenticate = function (username, password, realm, servi
...
         return Promise.resolve();
 
-    return request.post(this.serverAddress + '/identity/logout', {
-        headers: {
-            host: this.hostname
-        },
-        form: {
-            subjectid: sessionId
+    var headers = {};
+    console.log('Logout cookie: ' + cookieName)
```
Comment (zoltantarcsay) [inline lib/openam.js]: please remove this console.log statement

File: lib/agent.js [hunk 2]
```diff
@@ -163,6 +165,22 @@ function PolicyAgent(options) {
...
+ * Builds realm path used by /json endpoints: for details see https://backstage.forgerock.com/docs/am/6/dev-guide/#sec-rest-realms
+ */
+
+ function buildRealmPath(realm) {
+    var result = '';
+        var subrealms = realm.split('/');
+        subrealms[0] = 'root';
+        subrealms.forEach(function(r) {
+            if (r.length > 0) 
+                result += '/realms/' + r;
```
Comment (zoltantarcsay) [inline lib/agent.js]: is this backwards compatible...?

Comment (zoltantarcsay) [reply lib/agent.js]: looks like it's only valid in AM 5.0 or newer

File: lib/openam.js [hunk 3]
```diff
@@ -214,14 +224,11 @@ OpenAMClient.prototype.validateAccessToken = function (accessToken, realm) {
  */
 OpenAMClient.prototype.getProfile = function (userId, realm, sessionId, cookieName) {
     //noinspection JSValidateTypes
-    return request.get(this.serverAddress + '/json/users/' + userId, {
+    return request.get(this.serverAddress + '/json' + realm + '/users/' + userId, {
```
Comment (zoltantarcsay) [inline lib/openam.js]: realm should have a default value of `/`, otherwise this will become `/jsonundefined/users`

Comment (zoltantarcsay) [reply lib/openam.js]: (actually it's probably always called with a value...)

File: lib/openam.js [hunk 4]
```diff
@@ -149,15 +156,18 @@ OpenAMClient.prototype.getCDSSOUrl = function (target, provider) {
  * to have the REST calls for policy evaluation privilege in OpenAM.
  *
  * @param {object} params Policy request params {@see https://backstage.forgerock.com/#!/docs/openam/current/dev-guide#rest-api-authz-policy-decisions}
+ * @param {string} [realm=/]
  * @param {string} sessionId OpenAM session ID
  * @param {string} cookieName OpenAM session cookie name
  * @return {Promise} Policy decision response
  */
-OpenAMClient.prototype.getPolicyDecision = function (params, sessionId, cookieName) {
+OpenAMClient.prototype.getPolicyDecision = function (params, realm, sessionId, cookieName) {
```
Comment (zoltantarcsay) [inline lib/openam.js]: please reorder the arguments to keep this public method backwards compatible; `realm` should be the last argument and it should have a default value of `/`

File: lib/openam.js [hunk 5]
```diff
@@ -214,14 +224,11 @@ OpenAMClient.prototype.validateAccessToken = function (accessToken, realm) {
...
 OpenAMClient.prototype.getProfile = function (userId, realm, sessionId, cookieName) {
     //noinspection JSValidateTypes
-    return request.get(this.serverAddress + '/json/users/' + userId, {
+    return request.get(this.serverAddress + '/json' + realm + '/users/' + userId, {
         headers: {
             host: this.hostname,
             cookie: cookieName + '=' + sessionId
         },
-        qs: {
-            realm: realm
```
Comment (zoltantarcsay) [inline lib/openam.js]: is this not supported in AM 6 anymore? (I know it's not in the docs...)

File: lib/agent.js (hunk 2 above)
Comment (zoltantarcsay) [reply lib/agent.js]: if AM < 5 and AM > 5 don't support the same API signature, the agent should try to use the correct format based on the AM version to keep backwards compatibility; i.e. check the server version somehow and then decide how to handle the realms

File: lib/agent.js [hunk 6]
```diff
@@ -163,6 +165,22 @@ function PolicyAgent(options) {
     });
 }
 
+/** 
+ * Builds realm path used by /json endpoints: for details see https://backstage.forgerock.com/docs/am/6/dev-guide/#sec-rest-realms
+ */
+
+ function buildRealmPath(realm) {
```
Comment (zoltantarcsay) [inline lib/agent.js]: this function should have a simple unit test

**Comment by zoltantarcsay** [comment]
The TravisCI build fails on old Node.js versions because of a dependency which uses `const`. Since there is no easy way of fixing that, I'll have to remove support for Node.js 0.x

File: lib/agent.js (hunk 2 above)
Comment (aldaris) [reply lib/agent.js]: The backwards compatible way to approach this would be to always specify the realm as a query parameter (rather than as path)

Comment (mdetko) [reply lib/agent.js]: Does it really have to be backward compatible? You could say use version up to 4.0.1 with older version of AM and new one with AM 5 or newer.

File: lib/openam.js (hunk 3 above)
Comment (mdetko) [reply lib/openam.js]: intention was it is always called with value, see agent.js

File: lib/agent.js (hunk 2 above)
Comment (zoltantarcsay) [reply lib/agent.js]: It does, since I'm using it with AM 13 :)

**Pull request merged**